### PR TITLE
Initialize projects using Astro runtime images

### PR DESCRIPTION
**Pull request by AmFlint**
## Description

With the recent additions of Astro runtimes, we want to allow users to initialize new airflow projects using the Astro runtime as a base image instead of the Astronomer Certified Airflow images.

By default, we want to initialize a project with Astro runtime images, but still allow users to use Astronomer Certified images by providing a flag like:
```bash
# this will initialize a project using Astronomer Certified Airflow images, which is the behaviour in current releases 0.28.* and below.
astro dev init --use-astronomer-certified
```

**This PR should not be merged before Astro Runtime is supported on all components of the platform (Houston etc...)**.

## 🎟 Issue(s)

https://github.com/astronomer/issues/issues/4253

## 🧪 Functional Testing

To test this feature:
- create a new directory 
- Initalize a new project with default latest astronomer runtime tag/version:
```bash
# this should initialize a project using the latest "stable" Astro Runtime tag
astro dev init

# you can check it by looking at the content of the generated Dockerfile
cat Dockerfile
# image should be astro-runtime:whichever version is the latest stable one
```

You can also try to generate a project using a specific airflow version:
```bash
astro dev init -v 2.2.0
```

Also, try and generate a project using Astronomer Certified images:
```bash
astro dev init --use-astronomer-certified

# with given airflow version
astro dev init --use-astronomer-certified -v 2.0.0
```

## 📋 Checklist

- [x] Rebased from the main (or release if patching) branch (before testing)
- [x] Ran `make test` before taking out of draft
- [x] Added/updated applicable tests
- [ ] Tested against [Houston-API](https://github.com/astronomer/houston-api/) and [Astronomer](https://github.com/astronomer/astronomer/) (if necessary).
- [ ] Communicated to/tagged owners of respective clients potentially impacted by these changes.
- [ ] Updated any related [documentation](https://github.com/astronomer/docs/)

### CODE REVIEW

**Review by neel-astro — COMMENTED** [review]
overall LGTM 👍 , have added few minor comments

File: airflow_versions/airflow_versions.go
Comment (AmFlint) [reply airflow_versions/airflow_versions.go:71]: I went to read the cloud-cli code for this same feature, they look at the runtime version (5.0.0 which is in an `alpha` channel for now and can't be used for production purposes).

Just like I do, I just added another check for the runtime version's release channel to only allow users to use stable releases.

Anyway, we probably won't merge this PR into this repository, since we'll do the convergence soon, and I'll probably just submit a PR to allow AC images and default to AC images if context is software (CLI logged into a gen1 cluster).

But I guess that most of this PR will be dropped since cloud-cli already developped this feature.

File: airflow/container.go
Comment (neel-astro) [inline airflow/container.go:138]: any reason to move away from image labels to parsing to the dockerfile? I feel picking version from image label is more reliable compared to parsing dockerfile

File: airflow_versions/airflow_versions.go
Comment (neel-astro) [inline airflow_versions/airflow_versions.go:74]:
```suggestion
		if err == nil && nextVersion.GreaterThan(latestVersion) {
			latestVersion = nextVersion
		}
```

**Review by neel-astro — COMMENTED** [review]
overall looks fine, have added few minor comments

File: airflow/container.go
Comment (AmFlint) [reply airflow/container.go:138]: Why is it more reliable ? The Airflow version is added to 2 places:
- On the actual tag of the image, which can be found by simply reading from a file, which is always present inside an astronomer/airflow project (Dockerfile)
- The label on the image, which needs to be first pulled, then docker inspect.

Also, I made the decision of moving to reading Dockerfile since runtime code in Cloud-cli already uses this approach, and I wanted to use a consistent approach for both AC and Runtimes.

Comment (neel-astro) [reply airflow/container.go:138]: > Why is it more reliable? 

Because users can use custom images built on top of AC or Runtime image, in which case it is not necessary for users will carry forward the same version tag, but those images will inherit label from the base image.

> The label on the image, which needs to be first pulled, then docker inspect

We don't need to separately pull the image in this case, since we are already building the image from Dockerfile beforehand.

Comment (AmFlint) [reply airflow/container.go:138]: OK I see, in that case I'll use labels since customer can use a custom image.

File: airflow/airflow.go
Comment (andriisoldatenko) [inline airflow/airflow.go:25]:
```suggestion
	triggererAllowedRuntimeVersion = "4.2.5"
```

File: airflow/container.go
Comment (andriisoldatenko) [reply airflow/container.go:138]: @AmFlint  / @neel-astro may be it can help https://github.com/astronomer/astro-runtime/blob/main/image/Dockerfile#L31

Comment (AmFlint) [reply airflow/container.go:138]: I'll take a look at this to see if it's reliable for runtimes too

File: airflow/airflow.go
Comment (AmFlint) [reply airflow/airflow.go:25]: I think the triggerer resource was introduced in runtime v 4.0.0

Comment (andriisoldatenko) [reply airflow/airflow.go:25]: it's less important, Gen1 works only starting 4.2.5.

Comment (andriisoldatenko) [reply airflow/airflow.go:25]: https://github.com/astronomer/houston-api/blob/main/src/lib/constants/index.js#L209 i still need to update it here after i merge and release https://github.com/astronomer/astro-runtime/pull/235

File: airflow/container.go
Comment (AmFlint) [reply airflow/container.go]: Added for both AC and Runtimes

**Review by neel-astro — CHANGES_REQUESTED** [review]
looks fine apart from one minor comment, also some tests are failing which you might want to have a look at

File: airflow/podman.go
Comment (AmFlint) [reply airflow/podman.go:351]: Local testing, I fixed it sorry about that

**Review by neel-astro — APPROVED** [review]
lgtm

**Review by neel-astro — APPROVED** [review]

**Comment by andriisoldatenko** [comment]
> By default, we want to initialize a project with Astro runtime images

Yes

Regarding:
>By default, we want to initialize a project with Astro runtime images
I think you need to check feature flag ` "astroRuntimeEnabled": true` otherwise default to AC 

```
{
  "data": {
    "appConfig": {
      "version": "0.27.1-103-g86f6bfb7-build38298",
      "baseDomain": "dev.astrodev.io",
      "smtpConfigured": true,
      "configureDagDeployment": true,
      "nfsMountDagDeployment": true,
      "manualReleaseNames": false,
      "prometheusEnabled": true,
      "elasticSearchEnabled": true,
      "hardDeleteDeployment": true,
      "manualNamespaceNames": false,
      "appSubName": "Astronomer Software",
      "featureFlags": {
        "manualReleaseNames": false,
        "configureDagDeployment": true,
        "nfsMountDagDeployment": true,
        "hardDeleteDeployment": true,
        "manualNamespaceNames": false,
        "namespaceFreeFormEntry": false,
        "kibanaUIEnabled": true,
        "grafanaUIEnabled": true,
        "triggererEnabled": false,
        "sysAdminScalabilityImprovementsEnabled": true,
        "allowedSystemLevelDomains": [
          "astronomer.io"
        ],
        "allowedSystemLevelDomainsEnabled": true,
        "gitSyncDagDeployment": true,
        "idpGroupsEnabled": true,
        "astroRuntimeEnabled": true
      },
     ....
        "statusPage": null
      },
      "__typename": "AppConfig"
    }
  }
}
``` 

wdyt @cmarteepants ?

**Comment by andriisoldatenko** [comment]
@AmFlint pls fix conflicts / and add more tests.

**Comment by AmFlint** [comment]
I can add the logic to check for the astroRuntimeEnabled feature flag if you want.

In the converged cli, we removed all logic coupled to Houston in dev commands, since it's used to develop locally, we don't want to tie it to the Astronomer Software.

However, in this CLI, it's only used by Software users, so I think it makes sense to check this feature flag.

**Comment by andriisoldatenko** [comment]
> this

correct, i'm worried if customer don't want to enable runtime and cli will automatically init runtime based image, and than he can deploy it by mistake.

**Review by andriisoldatenko — APPROVED** [review]

**Comment by andriisoldatenko** [comment]
@AmFlint  / @neel-astro 

practically LGTM:
```
➜  t_20220413_130921 astro-cli dev init
Initialized empty astronomer project in /Users/andrii/workspace/src/github.com/astronomer/t_20220413_130921
➜  t_20220413_130921 cat Dockerfile
FROM quay.io/astronomer/astro-runtime:4.2.5%
➜  t_20220413_130921
```

**Pull request merged**
